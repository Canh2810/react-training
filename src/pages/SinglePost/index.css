.single-post {
  padding: 32px 209px 80px;
}

.single-post h1 {
  margin-top: 16px;
  margin-bottom: 20px;
}

.single-post__short-info,
.single-post__short-info-author {
  display: flex;
  align-items: center;
}

.single-post__short-info {
  gap: 24px;
  margin-bottom: 32px;
}

.single-post__short-info-author {
  gap: 8px;
}

.single-post__short-info-avatar {
  width: 28px;
  height: 28px;
  border-radius: 50%;
}

.single-post__img {
  width: 100%;
  height: auto;
  border-radius: 12px;
  margin-bottom: 32px;
}

.single-post article h2 {
  margin-top: 32px;
  margin-bottom: 16px;
}

.single-post__quote {
  padding: 32px;
  border-radius: 12px;
  background-color: var(--background-tertiary-color);
  margin: 32px 0;
}

.single-post__quote-content {
  font-style: italic;
}
