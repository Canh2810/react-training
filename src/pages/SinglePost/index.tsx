// Libs
import { memo } from 'react'
import { useParams } from 'react-router-dom'

// Components
import {
  Advertisement,
  Badge,
  LoadingIndicator,
  Typography,
  SearchResults,
} from '@components'

// Types
import { TypoVariants } from '@types'

//Style
import './index.css'
import { useDebounce, useFetchPost, useFetchPosts } from '@hooks'
import { usePostsStore, useQueryStore } from '@stores'

const SinglePost = () => {
  const param = useParams()
  const { id } = param || {}
  const { posts } = usePostsStore((state) => state)
  const { query } = useQueryStore((state) => state)
  const {
    data,
    isLoading: isLoadingFetchPost,
    isError: isErrorFetchPost,
    error: errorFetchPost,
  } = useFetchPost(id || '')

  const debouncedValue = useDebounce<string>(query)
  const isEnableFetchListPosts = query.length > 0

  const {
    isFetching: isLoadingFetchListPosts,
    isError: isErrorFetchListPosts,
    error: errorFetchListPosts,
  } = useFetchPosts('', debouncedValue, isEnableFetchListPosts)

  const { category, title, author, date, imageURL } = data || {}
  const { avatar, username } = author || {}

  const isLoading = isLoadingFetchPost || isLoadingFetchListPosts
  const isError = isErrorFetchPost || isErrorFetchListPosts
  const errorMessage = errorFetchPost?.message || errorFetchListPosts?.message

  return (
    <article className="single-post">
      {isError && <Typography>{errorMessage}</Typography>}
      {debouncedValue.length > 0 ? (
        <SearchResults posts={posts} />
      ) : (
        <>
          <Badge title={category} />
          <Typography component="h1" variant={TypoVariants.HeadingExtraLarge}>
            {title}
          </Typography>
          <div className="single-post__short-info">
            <div className="single-post__short-info-author">
              <img
                className="single-post__short-info-avatar"
                src={avatar}
                alt={username}
              />
              <Typography variant={TypoVariants.TextExtraSmall}>
                {username}
              </Typography>
            </div>
            <Typography variant={TypoVariants.TextExtraSmall}>
              {date}
            </Typography>
          </div>
          <img src={imageURL} alt={title} className="single-post__img" />
          <article>
            <Typography variant={TypoVariants.TextLarge}>
              Traveling is an enriching experience that opens up new horizons,
              exposes us to different cultures, and creates memories that last a
              lifetime. However, traveling can also be stressful and
              overwhelming, especially if you don't plan and prepare adequately.
              In this blog article, we'll explore tips and tricks for a
              memorable journey and how to make the most of your travels. One of
              the most rewarding aspects of traveling is immersing yourself in
              the local culture and customs. This includes trying local cuisine,
              attending cultural events and festivals, and interacting with
              locals. Learning a few phrases in the local language can also go a
              long way in making connections and showing respect.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Research Your Destination
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              Before embarking on your journey, take the time to research your
              destination. This includes understanding the local culture,
              customs, and laws, as well as identifying top attractions,
              restaurants, and accommodations. Doing so will help you navigate
              your destination with confidence and avoid any cultural faux pas.
              Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do
              eiusmod tempor incididunt ut labore et dolore magna aliqua. In
              hendrerit gravida rutrum quisque non tellus orci ac auctor. Mi
              ipsum faucibus vitae aliquet nec ullamcorper sit amet. Aenean
              euismod elementum nisi quis eleifend quam adipiscing vitae.
              Viverra adipiscing at in tellus.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Plan Your Itinerary
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              While it's essential to leave room for spontaneity and unexpected
              adventures, having a rough itinerary can help you make the most of
              your time and budget. Identify the must-see sights and experiences
              and prioritize them according to your interests and preferences.
              This will help you avoid overscheduling and ensure that you have
              time to relax and enjoy your journey. Vitae sapien pellentesque
              habitant morbi tristique. Luctus venenatis lectus magna fringilla.
              Nec ullamcorper sit amet risus nullam eget felis. Tincidunt arcu
              non sodales neque sodales ut etiam sit amet.
            </Typography>
            <div className="single-post__quote">
              <Typography
                variant={TypoVariants.TexExtraLarge}
                className="single-post__quote-content"
              >
                “ Traveling can expose you to new environments and potential
                health risks, so it's crucial to take precautions to stay safe
                and healthy. ”
              </Typography>
            </div>
            <img
              src="https://plus.unsplash.com/premium_photo-1666963323736-5ee1c16ef19d?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8MTN8fG5hdHVyYWx8ZW58MHx8MHx8fDA%3D&auto=format&fit=crop&w=600&q=60"
              alt="Plan Your Itinerary"
              className="single-post__img"
            />
          </article>
          <Advertisement />
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Pack Lightly and Smartly
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              Packing can be a daunting task, but with some careful planning and
              smart choices, you can pack light and efficiently. Start by making
              a packing list and sticking to it, focusing on versatile and
              comfortable clothing that can be mixed and matched. Invest in
              quality luggage and packing organizers to maximize space and
              minimize wrinkles.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Stay Safe and Healthy
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              Traveling can expose you to new environments and potential health
              risks, so it's crucial to take precautions to stay safe and
              healthy. This includes researching any required vaccinations or
              medications, staying hydrated, washing your hands frequently, and
              using sunscreen and insect repellent. It's also essential to keep
              your valuables safe and secure and to be aware of your
              surroundings at all times.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Immerse Yourself in the Local Culture
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              One of the most rewarding aspects of traveling is immersing
              yourself in the local culture and customs. This includes trying
              local cuisine, attending cultural events and festivals, and
              interacting with locals. Learning a few phrases in the local
              language can also go a long way in making connections and showing
              respect.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Capture Memories
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              Finally, don't forget to capture memories of your journey. Whether
              it's through photographs, journaling, or souvenirs, preserving the
              moments and experiences of your travels can bring joy and
              nostalgia for years to come. However, it's also essential to be
              present in the moment and not let technology distract you from the
              beauty of your surroundings.
            </Typography>
          </article>
          <article>
            <Typography component="h2" variant={TypoVariants.HeadingLarge}>
              Conclusion:
            </Typography>
            <Typography variant={TypoVariants.TextLarge}>
              Traveling is an art form that requires a blend of planning,
              preparation, and spontaneity. By following these tips and tricks,
              you can make the most of your journey and create memories that
              last a lifetime. So pack your bags, embrace the adventure, and
              enjoy the ride.
            </Typography>
          </article>
        </>
      )}
      {isLoading && <LoadingIndicator />}
    </article>
  )
}

export default memo(SinglePost)
