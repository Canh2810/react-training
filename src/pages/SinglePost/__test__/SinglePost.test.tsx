import { render, screen, waitFor } from '@testing-library/react'
import SinglePost from '..'
import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { MemoryRouter } from 'react-router-dom'
import { POSTS_MOCK, POST_MOCK, THEMES } from '@constants'
import { usePostsStore, useQueryStore, useThemeStore } from '@stores'

const queryClient = new QueryClient()

const setup = () =>
  render(
    <QueryClientProvider client={queryClient}>
      <SinglePost />
    </QueryClientProvider>,
    { wrapper: MemoryRouter },
  )

const mockUseFetchPosts = jest.fn()
const mockUseFetchPost = jest.fn()
const mockUseDebounce = jest.fn()

jest.mock('@hooks', () => ({
  ...jest.requireActual('@hooks'),
  useFetchPosts: () => mockUseFetchPosts,
  useFetchPost: () => mockUseFetchPost,
  useDebounce: () => mockUseDebounce,
}))

describe('SinglePost page', () => {
  beforeEach(() => {
    mockUseFetchPosts.mockReturnValue({
      isLoading: false,
      isError: false,
      error: null,
    }),
      mockUseFetchPost.mockReturnValue({
        data: POST_MOCK,
        isLoading: false,
        isError: false,
        error: null,
      }),
      mockUseDebounce.mockReturnValue(''),
      useThemeStore.setState({
        theme: THEMES.LIGHT,
      }),
      usePostsStore.setState({
        posts: POSTS_MOCK,
      }),
      useQueryStore.setState({
        query: '',
      })
  })

  it('should render correctly', () => {
    const { container } = setup()
    expect(container).toMatchSnapshot()
  })

  it('should show error message if there is an error while fetching post', () => {
    mockUseFetchPost.mockReturnValue({
      data: [],
      isLoading: false,
      isError: true,
      error: { message: 'Error fetching post' },
    })

    setup()
    waitFor(() => {
      expect(screen.getByText('Error fetching post')).toBeInTheDocument()
    })
  })

  it('should show loading indicator while fetching data', () => {
    mockUseFetchPost.mockReturnValue({
      data: POST_MOCK,
      isLoading: true,
      isError: false,
      error: null,
    })

    setup()
    waitFor(() => {
      expect(screen.getByTestId('loading-indicator')).toBeInTheDocument()
    })
  })

  it('should render search results when there is a query', () => {
    mockUseDebounce.mockReturnValue('abc')

    waitFor(() => {
      expect(screen.getByTestId('No Result Found')).toBeInTheDocument()
    })
  })
})
