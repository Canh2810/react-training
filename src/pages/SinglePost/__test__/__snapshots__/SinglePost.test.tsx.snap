// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SinglePost page should render correctly 1`] = `
<div>
  <article
    class="single-post"
  >
    <span
      class="badge"
      style="background-color: rgb(75, 107, 251); color: rgb(255, 255, 255);"
    />
    <h1
      class=""
      style="font-family: work-sans; font-size: 36px; font-weight: 600; line-height: 40px; color: rgb(59, 60, 74);"
    />
    <div
      class="single-post__short-info"
    >
      <div
        class="single-post__short-info-author"
      >
        <img
          class="single-post__short-info-avatar"
        />
        <p
          class=""
          style="font-family: inter; font-size: 14px; font-weight: 400; line-height: 20px; color: rgb(59, 60, 74);"
        />
      </div>
      <p
        class=""
        style="font-family: inter; font-size: 14px; font-weight: 400; line-height: 20px; color: rgb(59, 60, 74);"
      />
    </div>
    <img
      class="single-post__img"
    />
    <article>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Traveling is an enriching experience that opens up new horizons, exposes us to different cultures, and creates memories that last a lifetime. However, traveling can also be stressful and overwhelming, especially if you don't plan and prepare adequately. In this blog article, we'll explore tips and tricks for a memorable journey and how to make the most of your travels. One of the most rewarding aspects of traveling is immersing yourself in the local culture and customs. This includes trying local cuisine, attending cultural events and festivals, and interacting with locals. Learning a few phrases in the local language can also go a long way in making connections and showing respect.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Research Your Destination
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Before embarking on your journey, take the time to research your destination. This includes understanding the local culture, customs, and laws, as well as identifying top attractions, restaurants, and accommodations. Doing so will help you navigate your destination with confidence and avoid any cultural faux pas. Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. In hendrerit gravida rutrum quisque non tellus orci ac auctor. Mi ipsum faucibus vitae aliquet nec ullamcorper sit amet. Aenean euismod elementum nisi quis eleifend quam adipiscing vitae. Viverra adipiscing at in tellus.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Plan Your Itinerary
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        While it's essential to leave room for spontaneity and unexpected adventures, having a rough itinerary can help you make the most of your time and budget. Identify the must-see sights and experiences and prioritize them according to your interests and preferences. This will help you avoid overscheduling and ensure that you have time to relax and enjoy your journey. Vitae sapien pellentesque habitant morbi tristique. Luctus venenatis lectus magna fringilla. Nec ullamcorper sit amet risus nullam eget felis. Tincidunt arcu non sodales neque sodales ut etiam sit amet.
      </p>
      <div
        class="single-post__quote"
      >
        <p
          class="single-post__quote-content"
          style="font-family: source-serif-pro; font-size: 24px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
        >
          “ Traveling can expose you to new environments and potential health risks, so it's crucial to take precautions to stay safe and healthy. ”
        </p>
      </div>
      <img
        alt="Plan Your Itinerary"
        class="single-post__img"
        src="https://plus.unsplash.com/premium_photo-1666963323736-5ee1c16ef19d?ixlib=rb-4.0.3&ixid=M3wxMjA3fDB8MHxzZWFyY2h8MTN8fG5hdHVyYWx8ZW58MHx8MHx8fDA%3D&auto=format&fit=crop&w=600&q=60"
      />
    </article>
    <div
      class="advertisement"
    >
      <p
        class=""
        style="font-family: inter; font-size: 14px; font-weight: 400; line-height: 20px; color: rgb(59, 60, 74);"
      >
        Advertisement
      </p>
      <p
        class=""
        style="font-family: work-sans; font-size: 20px; font-weight: 600; line-height: 28px; color: rgb(105, 106, 117);"
      >
        You can place ads
      </p>
      <p
        class=""
        style="font-family: work-sans; font-size: 18px; font-weight: 400; line-height: 26px; color: rgb(59, 60, 74);"
      >
        750x100
      </p>
    </div>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Pack Lightly and Smartly
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Packing can be a daunting task, but with some careful planning and smart choices, you can pack light and efficiently. Start by making a packing list and sticking to it, focusing on versatile and comfortable clothing that can be mixed and matched. Invest in quality luggage and packing organizers to maximize space and minimize wrinkles.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Stay Safe and Healthy
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Traveling can expose you to new environments and potential health risks, so it's crucial to take precautions to stay safe and healthy. This includes researching any required vaccinations or medications, staying hydrated, washing your hands frequently, and using sunscreen and insect repellent. It's also essential to keep your valuables safe and secure and to be aware of your surroundings at all times.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Immerse Yourself in the Local Culture
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        One of the most rewarding aspects of traveling is immersing yourself in the local culture and customs. This includes trying local cuisine, attending cultural events and festivals, and interacting with locals. Learning a few phrases in the local language can also go a long way in making connections and showing respect.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Capture Memories
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Finally, don't forget to capture memories of your journey. Whether it's through photographs, journaling, or souvenirs, preserving the moments and experiences of your travels can bring joy and nostalgia for years to come. However, it's also essential to be present in the moment and not let technology distract you from the beauty of your surroundings.
      </p>
    </article>
    <article>
      <h2
        class=""
        style="font-family: work-sans; font-size: 24px; font-weight: 600; line-height: 28px; color: rgb(59, 60, 74);"
      >
        Conclusion:
      </h2>
      <p
        class=""
        style="font-family: source-serif-pro; font-size: 20px; font-weight: 400; line-height: 32px; color: rgb(59, 60, 74);"
      >
        Traveling is an art form that requires a blend of planning, preparation, and spontaneity. By following these tips and tricks, you can make the most of your journey and create memories that last a lifetime. So pack your bags, embrace the adventure, and enjoy the ride.
      </p>
    </article>
  </article>
</div>
`;
